### DISCUSSION TITLE
Raspberry PI

### DISCUSSION THREAD

**Discussion by meeximum** [rathole-org/rathole#233] [Q&A] [answered]
Hi,

I want to use rathole on my raspberry pi, which asset is right for me?

kind regards
Meex

**Comment by yujqiao** [comment, marked as answer]
Depends on the architecture of the very model.

One of the followings should work:

- `arm-unknown-linux-musleabi` 
- `arm-unknown-linux-musleabihf`
- `armv7-unknown-linux-musleabihf` for recent PIs

> For a Raspberry Pi 4 (and also for most other recent Raspberry Pi boards), we can use the armv7-unknown-linux-musleabihf target.
Roughly, the target components have the following meaning:
armv7 : the architecture to use for the target processor, in our case ARM v7 (even though [Raspberry Pi 4 actually supports up to ARM v8](https://www.raspberrypi.org/products/raspberry-pi-4-model-b/specifications/))
unknown : the sub-architecture to use; in our case this just means the [default option](https://clang.llvm.org/docs/CrossCompilation.html#target-triple)
linux : the target Operating System
musleabihf : the ABI to target; musl means that it relies on muslc for some functionality at runtime; hf means hard float, that is that the architecture supports hardware floating-point operations
(thanks to https://medium.com/swlh/compiling-rust-for-raspberry-pi-arm-922b55dbb050)

**Reply by meeximum** [reply]
thx for your quick and detailed reply!

**Comment by zdeneksvarc** [comment]
Thank you guys. This Dockerfile works for RPi 4:

```
FROM alpine:3.15 as extractor
RUN apk add --no-cache curl unzip
WORKDIR /download
RUN curl -sL https://github.com/rapiz1/rathole/releases/download/v0.5.0/rathole-armv7-unknown-linux-musleabihf.zip -o rathole.zip \
    && unzip rathole.zip

FROM gcr.io/distroless/cc-debian12
WORKDIR /app
COPY --from=extractor /download/rathole .
USER 1000:1000
ENTRYPOINT ["./rathole"]
```
